// @react-keycloak/web/dist/lib/react-keycloak-web.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var core = require('@react-keycloak/core');
var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _assertThisInitialized = _interopDefault(require('@babel/runtime/helpers/assertThisInitialized'));
var _inheritsLoose = _interopDefault(require('@babel/runtime/helpers/inheritsLoose'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var React = require('react');
var hoistStatics = _interopDefault(require('hoist-non-react-statics'));

var reactKeycloakWebContext = core.createReactKeycloakContext();

var reactKeycloakProvider = core.createReactKeycloakProvider(reactKeycloakWebContext);

var getDisplayName = function getDisplayName(name) {
  return "WithKeycloak(" + name + ")";
};

function withKeycloak(WrappedComponent) {
  var Context = reactKeycloakWebContext;
  var displayName = getDisplayName(WrappedComponent.displayName || WrappedComponent.name);

  var WithKeycloakComponent = /*#__PURE__*/function (_React$PureComponent) {
    _inheritsLoose(WithKeycloakComponent, _React$PureComponent);

    function WithKeycloakComponent() {
      var _this;

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = _React$PureComponent.call.apply(_React$PureComponent, [this].concat(args)) || this;

      _defineProperty(_assertThisInitialized(_this), "renderWrappedComponent", function (_ref) {
        var initialized = _ref.initialized,
            keycloak = _ref.keycloak;
        return /*#__PURE__*/React.createElement(WrappedComponent, _extends({}, _this.props, {
          keycloak: keycloak,
          keycloakInitialized: initialized
        }));
      });

      return _this;
    }

    var _proto = WithKeycloakComponent.prototype;

    _proto.render = function render() {
      return /*#__PURE__*/React.createElement(Context.Consumer, null, this.renderWrappedComponent);
    };

    return WithKeycloakComponent;
  }(React.PureComponent);

  WithKeycloakComponent.WrappedComponent = WrappedComponent;
  WithKeycloakComponent.displayName = displayName;
  return hoistStatics(WithKeycloakComponent, WrappedComponent);
}

function useKeycloak() {
  var _useContext = React.useContext(reactKeycloakWebContext),
      initialized = _useContext.initialized,
      keycloak = _useContext.keycloak;

  return Object.assign([keycloak, initialized], {
    initialized: initialized,
    keycloak: keycloak
  });
}

exports.KeycloakProvider = reactKeycloakProvider;
exports.useKeycloak = useKeycloak;
exports.withKeycloak = withKeycloak;
